routing/index_router.cpp

```diff
@@ -558,29 +592,39 @@ bool IndexRouter::FindBestSegment(m2::PointD const & point, m2::PointD const & d
   vector<pair<Edge, Junction>> candidates;
   m_roadGraph.FindClosestEdges(point, kMaxRoadCandidates, candidates);
 
-  auto const getSegmentByEdge = [&numMwmId](Edge const & edge){
-      return Segment(numMwmId, edge.GetFeatureId().m_index, edge.GetSegId(), edge.IsForward());
+  auto const getSegmentByEdge = [&numMwmId](Edge const & edge, bool isForward) {
+    return Segment(numMwmId, edge.GetFeatureId().m_index, edge.GetSegId(), isForward);
   };
 
   // Getting rid of knowingly bad candidates.
   my::EraseIf(candidates, [&](pair<Edge, Junction> const & p){
     Edge const & edge = p.first;
-    return edge.GetFeatureId().m_mwmId != mwmId || IsDeadEnd(getSegmentByEdge(edge), isOutgoing, worldGraph);
+    return edge.GetFeatureId().m_mwmId != mwmId ||
+           IsDeadEnd(getSegmentByEdge(edge, edge.IsForward()), isOutgoing, worldGraph);
   });
 
   if (candidates.empty())
     return false;
 
-  BestEdgeComparator bestEdgeComparator(point, direction);
-  Edge & bestEdge = candidates[0].first;
+  BestEdgeComparator bestEdgeComparator(point, direction, [this](FeatureID const & featureId) {
+    return this->IsOneWayFeature(featureId);
+  });
+  Edge bestEdge = candidates[0].first;
   for (size_t i = 1; i < candidates.size(); ++i)
   {
     Edge const & edge = candidates[i].first;
     if (bestEdgeComparator.Compare(edge, bestEdge))
       bestEdge = edge;
   }
 
-  bestSegment = getSegmentByEdge(bestEdge);
+  bestSegmentIsAlmostConformDirection =
+      bestEdgeComparator.IsDirectionValid() && bestEdgeComparator.IsAlmostConformed(bestEdge);
+
+  bool bestSegmentForward = true;
+  if (bestEdgeComparator.IsDirectionValid())
+    bestEdgeComparator.IsAlmostCollinear(bestEdge, bestSegmentForward);
```

Comment:
IsAlmostCollinear sounds like a predicate, and what's the purpose to call a predicate and do not use return value?